File: apps/desktop/frontend/src/services/storage.ts
// 本地文件系统存储服务 - 使用 .disk-rookie 目录
import { invoke } from '@tauri-apps/api/core'

/**
 * 读取存储文件
 */
export async function readStorageFile(filename: string): Promise<string> {
  try {
    return await invoke<string>('read_storage_file', { filename })
  } catch (error) {
    console.error(`读取文件失败 ${filename}:`, error)
    return ''
  }
}

/**
 * 写入存储文件
 */
export async function writeStorageFile(filename: string, content: string): Promise<void> {
  try {
    await invoke('write_storage_file', { filename, content })
  } catch (error) {
    console.error(`写入文件失败 ${filename}:`, error)
    throw error
  }
}

/**
 * 删除存储文件
 */
export async function deleteStorageFile(filename: string): Promise<void> {
  try {
    await invoke('delete_storage_file', { filename })
  } catch (error) {
    console.error(`删除文件失败 ${filename}:`, error)
    throw error
  }
}

/**
 * 列出存储文件
 */
export async function listStorageFiles(subdir?: string): Promise<string[]> {
  try {
    return await invoke<string[]>('list_storage_files', { subdir })
  } catch (error) {
    console.error('列出文件失败:', error)
    return []
  }
}

/**
 * 获取存储根目录路径
 */
export async function getStoragePath(): Promise<string> {
  try {
    return await invoke<string>('get_storage_path')
  } catch (error) {
    console.error('获取存储路径失败:', error)
    return ''
  }
}

/**
 * 读取 JSON 文件
 */
export async function readJSON<T>(filename: string, defaultValue: T): Promise<T> {
  try {
    const content = await readStorageFile(filename)
    if (!content) return defaultValue
    return JSON.parse(content) as T
  } catch (error) {
    console.error(`读取 JSON 失败 ${filename}:`, error)
    return defaultValue
  }
}

/**
 * 写入 JSON 文件
 */
export async function writeJSON(filename: string, data: unknown): Promise<void> {
  const content = JSON.stringify(data, null, 2)
  await writeStorageFile(filename, content)
}

File: apps/desktop/src-tauri/src/commands/storage.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;
use tauri::{AppHandle, Manager};

#[derive(Debug, Serialize, Deserialize)]
pub struct StorageResult {
    success: bool,
    message: String,
}

/// 获取存储根目录 (.disk-rookie)
fn get_storage_root(app: &AppHandle) -> Result<PathBuf, String> {
    let home_dir = app
        .path()
        .home_dir()
        .map_err(|e| format!("无法获取用户目录: {}", e))?;
    
    let storage_root = home_dir.join(".disk-rookie");
    
    // 确保目录存在
    if !storage_root.exists() {
        fs::create_dir_all(&storage_root)
            .map_err(|e| format!("创建存储目录失败: {}", e))?;
    }
    
    Ok(storage_root)
}

/// 读取文件
#[tauri::command]
pub async fn read_storage_file(app: AppHandle, filename: String) -> Result<String, String> {
    let storage_root = get_storage_root(&app)?;
    let file_path = storage_root.join(&filename);
    
    if !file_path.exists() {
        return Ok(String::new());
    }
    
    fs::read_to_string(&file_path)
        .map_err(|e| format!("读取文件失败 {}: {}", filename, e))
}

/// 写入文件
#[tauri::command]
pub async fn write_storage_file(
    app: AppHandle,
    filename: String,
    content: String,
) -> Result<StorageResult, String> {
    let storage_root = get_storage_root(&app)?;
    let file_path = storage_root.join(&filename);
    
    // 确保父目录存在
    if let Some(parent) = file_path.parent() {
        if !parent.exists() {
            fs::create_dir_all(parent)
                .map_err(|e| format!("创建目录失败: {}", e))?;
        }
    }
    
    fs::write(&file_path, content)
        .map_err(|e| format!("写入文件失败 {}: {}", filename, e))?;
    
    Ok(StorageResult {
        success: true,
        message: format!("文件已保存: {}", filename),
    })
}

/// 删除文件
#[tauri::command]
pub async fn delete_storage_file(app: AppHandle, filename: String) -> Result<StorageResult, String> {
    let storage_root = get_storage_root(&app)?;
    let file_path = storage_root.join(&filename);
    
    if file_path.exists() {
        fs::remove_file(&file_path)
            .map_err(|e| format!("删除文件失败 {}: {}", filename, e))?;
    }
    
    Ok(StorageResult {
        success: true,
        message: format!("文件已删除: {}", filename),
    })
}

/// 列出目录中的文件
#[tauri::command]
pub async fn list_storage_files(app: AppHandle, subdir: Option<String>) -> Result<Vec<String>, String> {
    let storage_root = get_storage_root(&app)?;
    let target_dir = if let Some(sub) = subdir {
        storage_root.join(sub)
    } else {
        storage_root
    };
    
    if !target_dir.exists() {
        return Ok(Vec::new());
    }
    
    let entries = fs::read_dir(&target_dir)
        .map_err(|e| format!("读取目录失败: {}", e))?;
    
    let mut files = Vec::new();
    for entry in entries {
        if let Ok(entry) = entry {
            if let Some(filename) = entry.file_name().to_str() {
                files.push(filename.to_string());
            }
        }
    }
    
    files.sort();
    Ok(files)
}

/// 获取存储根目录路径
#[tauri::command]
pub async fn get_storage_path(app: AppHandle) -> Result<String, String> {
    let storage_root = get_storage_root(&app)?;
    storage_root
        .to_str()
        .map(|s| s.to_string())
        .ok_or_else(|| "无法转换路径".to_string())
}
